### PR TITLE
Fix data races in hipcpu::detail::sync_queue

### PR DESCRIPTION
**Pull request by nazavode**
This PR fixes several data races detected by `TSAN` while using `hipCPU` as a dependency vendored by `hipSYCL`. Each commit addresses a specific data race while all of them boil down to unsynced accesses to both `sync_queue::_continue` and `sync_queue::_enqueued_operations` (for example, full `TSAN` log of the race fixed by 48eeeff is [here](https://gist.github.com/nazavode/728ff2260ed5ebd3e800a5b44cadaa8f)).

### CODE REVIEW

File: include/hipCPU/hip/detail/queue.hpp [hunk 1]
```diff
@@ -103,7 +103,12 @@ async_queue::~async_queue()
 {
   halt();
 
-  assert(_enqueued_operations.empty());
+  #if !defined(NDEBUG)
+  {
+    std::unique_lock<std::mutex> lock(_mutex);
+    assert(_enqueued_operations.empty());
```
Comment (illuhad) [inline include/hipCPU/hip/detail/queue.hpp]: Can you elaborate why this change is necessary? Because of `halt()`, all asynchronous operations should have finished and the threads rejoined by the time of the assertion. Am I missing something?

File: include/hipCPU/hip/detail/queue.hpp [hunk 2]
```diff
@@ -170,9 +179,11 @@ void async_queue::work()
     _is_idle = false;
     operation();
 
-    if(_enqueued_operations.empty())
-      _is_idle = true;
-
+    _is_idle = [this] {
+      std::lock_guard<std::mutex> lock(_mutex);
+      return _enqueued_operations.empty();
+    }();
```
Comment (illuhad) [inline include/hipCPU/hip/detail/queue.hpp]: I wonder if it might be clearer to just have 
```c++
{
  std::local_guard<std::mutex> lock(_mutex);
  _is_idle = _enqueued_operations.empty();
}
```
?

**Comment by illuhad** [comment]
Thank you! I have a couple of questions. Also, we will need to backport these changes into hipSYCL, which uses the same implementation for the background worker thread inside the runtime  in the file `async_worker.hpp`.
[We also need this functionality in hipSYCL for the CUDA/ROCm backends where hipCPU is not available]

File: include/hipCPU/hip/detail/queue.hpp (hunk 2 above)
Comment (nazavode) [reply include/hipCPU/hip/detail/queue.hpp]: Sure, you're right, I was wondering how to sync the access to `_enqueued_operations` only since `_is_idle` is atomic anyway and overlook the overly complicated solution I used :)

**Comment by nazavode** [comment]
Similar issues fixed in illuhad/hipSYCL#233: since I'm going to address your comments here, I think it could be better to iterate to an acceptable solution here and then backport the relevant changes to illuhad/hipSYCL#233.

File: include/hipCPU/hip/detail/queue.hpp (hunk 1 above)
Comment (nazavode) [reply include/hipCPU/hip/detail/queue.hpp]: Let me investigate this a bit more in depth, your point about joined threads ringed a bell...

Comment (nazavode) [reply include/hipCPU/hip/detail/queue.hpp]: You were totally right, that synced scope was totally useless, TSAN is still pretty happy without it. Just dropped the commit.

**Comment by illuhad** [comment]
Sorry for the delay, was busy with preparing IWOCL talks :) This looks good now, thank you!

**Comment by nazavode** [comment]
@illuhad looking forward to IWOCL :)
